From d13df2a31e5667e98251f34de0a7ee0d4cb344fc Mon Sep 17 00:00:00 2001
From: Atakan <ata.kaner03@gmail.com>
Date: Fri, 6 Dec 2024 12:30:45 +0300
Subject: [PATCH 1/4] sdm: Support Samsung FINGERPRINT_MASK DRM property

Allow Samsung devices to enable their HBM mode when a dim layer
is present.

Change-Id: I200d06d1e17dbad55715d4fc8e1ce812ff0f0ec9
---
 Android.bp                          |  8 +++++++
 libdrmutils/drm_interface.h         |  8 +++++++
 sde-drm/drm_atomic_req.cpp          |  3 +++
 sde-drm/drm_connector.cpp           | 18 +++++++++++++++
 sde-drm/drm_property.cpp            |  3 +++
 sde-drm/drm_property.h              |  3 +++
 sdm/libs/core/drm/hw_device_drm.cpp | 35 +++++++++++++++++++++++++++++
 sdm/libs/core/drm/hw_device_drm.h   |  3 +++
 8 files changed, 81 insertions(+)

diff --git a/Android.bp b/Android.bp
index 18cdebdc68..b96f7aee02 100644
--- a/Android.bp
+++ b/Android.bp
@@ -8,6 +8,7 @@ soong_config_module_type {
         "llvmsa",
         "gralloc4",
         "udfps",
+        "samsung_udfps",
         "default",
         "var3",
     ],
@@ -107,6 +108,13 @@ qtidisplay_cc_defaults {
                 enabled: true,
             },
         },
+        samsung_udfps: {
+            cflags: [
+                "-DCONNECTOR_PROP_UDFPS",
+                "-DSEC_FINGERPRINT_MASK",
+                "-DUDFPS_ZPOS",
+            ],
+        },
     },
 }
 
diff --git a/libdrmutils/drm_interface.h b/libdrmutils/drm_interface.h
index 8814e6b9cd..1291b9944b 100644
--- a/libdrmutils/drm_interface.h
+++ b/libdrmutils/drm_interface.h
@@ -505,6 +505,14 @@ enum struct DRMOps {
    * Arg: drmModeAtomicReq - Atomic request
    */
   RESET_PANEL_FEATURES,
+#ifdef CONNECTOR_PROP_UDFPS
+  /*
+   * Op: Sets fingerprint mask on this connector
+   * Arg: uint32_t - Connector ID
+   *      uint32_t - fingerprint mask
+   */
+  CONNECTOR_SET_FINGERPRINT_MASK,
+#endif
 };
 
 enum struct DRMRotation {
diff --git a/sde-drm/drm_atomic_req.cpp b/sde-drm/drm_atomic_req.cpp
index cdf77c1656..fceb7508b2 100644
--- a/sde-drm/drm_atomic_req.cpp
+++ b/sde-drm/drm_atomic_req.cpp
@@ -130,6 +130,9 @@ int DRMAtomicReq::Perform(DRMOps opcode, uint32_t obj_id, ...) {
     case DRMOps::CONNECTOR_SET_COLORSPACE:
     case DRMOps::CONNECTOR_SET_PANEL_MODE: 
     case DRMOps::CONNECTOR_SET_DYN_BIT_CLK:
+#ifdef CONNECTOR_PROP_UDFPS
+    case DRMOps::CONNECTOR_SET_FINGERPRINT_MASK:
+#endif
     case DRMOps::CONNECTOR_SET_DSC_MODE: {
       drm_mgr_->GetConnectorMgr()->Perform(opcode, obj_id, drm_atomic_req_, args);
     } break;
diff --git a/sde-drm/drm_connector.cpp b/sde-drm/drm_connector.cpp
index ba29c81b7c..a2c82316b9 100644
--- a/sde-drm/drm_connector.cpp
+++ b/sde-drm/drm_connector.cpp
@@ -1276,6 +1276,24 @@ void DRMConnector::Perform(DRMOps code, drmModeAtomicReq *req, va_list args) {
         DRM_LOGD("Connector %d: Setting compression mode %d", obj_id, drm_compression_mode);
       }
     } break;
+#ifdef CONNECTOR_PROP_UDFPS
+    case DRMOps::CONNECTOR_SET_FINGERPRINT_MASK: {
+      if (!prop_mgr_.IsPropertyAvailable(DRMProperty::FINGERPRINT_MASK)) {
+        DRM_LOGE("DRMConnector::%s: Connector %d: Fingerprint property is not available",
+                 __FUNCTION__, obj_id);
+        return;
+      }
+      uint32_t fingerprint_mask = va_arg(args, uint32_t);
+      uint32_t prop_id = prop_mgr_.GetPropertyId(DRMProperty::FINGERPRINT_MASK);
+      int ret = drmModeAtomicAddProperty(req, obj_id, prop_id, fingerprint_mask);
+      if (ret < 0) {
+        DRM_LOGE("AtomicAddProperty failed obj_id 0x%x, prop_id %d mode %d ret %d",
+                 obj_id, prop_id, fingerprint_mask, ret);
+      } else {
+        DRM_LOGD("Connector %d: Setting Fingerprint mode %d", obj_id, fingerprint_mask);
+      }
+    } break;
+#endif
 
     default:
       DRM_LOGE("Invalid opcode %d to set on connector %d", code, obj_id);
diff --git a/sde-drm/drm_property.cpp b/sde-drm/drm_property.cpp
index 5de330ce95..1f789f531f 100644
--- a/sde-drm/drm_property.cpp
+++ b/sde-drm/drm_property.cpp
@@ -180,6 +180,9 @@ DRMProperty DRMPropertyManager::GetPropertyEnum(const std::string &name) const {
   if (name == "dimming_bl_lut") { return DRMProperty::DIMMING_BL_LUT; }
   if (name == "dimming_dyn_ctrl") { return DRMProperty::DIMMING_DYN_CTRL; }
   if (name == "dimming_min_bl") { return DRMProperty::DIMMING_MIN_BL; }
+#ifdef CONNECTOR_PROP_UDFPS
+  if (name == "fingerprint_mask") { return DRMProperty::FINGERPRINT_MASK; }
+#endif
 
   return DRMProperty::INVALID;
 }
diff --git a/sde-drm/drm_property.h b/sde-drm/drm_property.h
index aee72beb60..08d278af34 100644
--- a/sde-drm/drm_property.h
+++ b/sde-drm/drm_property.h
@@ -185,6 +185,9 @@ enum struct DRMProperty {
   DIMMING_BL_LUT,
   DIMMING_DYN_CTRL,
   DIMMING_MIN_BL,
+#ifdef CONNECTOR_PROP_UDFPS
+  FINGERPRINT_MASK,
+#endif
 
   // Insert above
   MAX
diff --git a/sdm/libs/core/drm/hw_device_drm.cpp b/sdm/libs/core/drm/hw_device_drm.cpp
index 04cc73bfce..a1eee4fff6 100644
--- a/sdm/libs/core/drm/hw_device_drm.cpp
+++ b/sdm/libs/core/drm/hw_device_drm.cpp
@@ -1227,6 +1227,17 @@ DisplayError HWDeviceDRM::PowerOff(bool teardown, SyncPoints *sync_points) {
     pending_power_state_ = kPowerStateOff;
     return kErrorDeferred;
   }
+#ifdef SEC_FINGERPRINT_MASK
+  if (IsPrimaryDisplay()) {
+    if (current_mask_state_) {
+      current_mask_state_ = 0;
+      drm_atomic_intf_->Perform(DRMOps::CONNECTOR_SET_FINGERPRINT_MASK,
+          token_.conn_id, current_mask_state_);
+      DLOGV("Display:%d Setting Fingerprint inDisplay Layer property = %d",
+            display_id_, current_mask_state_);
+    }
+  }
+#endif
 
   ResetROI();
   int64_t retire_fence_fd = -1;
@@ -1392,6 +1403,30 @@ void HWDeviceDRM::SetupAtomic(Fence::ScopedRef &scoped_ref, HWLayersInfo *hw_lay
                        hw_layers_info->flags.geometry_changed;
   bool update_luts = hw_layers_info->updates_mask.test(kUpdateLuts);
 
+#ifdef SEC_FINGERPRINT_MASK
+  if (IsPrimaryDisplay()) {
+    uint32_t mask_state = 0;
+
+    for (uint32_t i = 0; i < hw_layer_count; i++) {
+      Layer &layer = hw_layer_info.hw_layers.at(i);
+      if (layer.flags.fod_pressed ||
+          (hw_layer_info.stack->flags.fod_pressed_present && i == hw_layer_count - 1)) {
+        mask_state = 1;
+        goto out;
+      }
+    }
+
+  out:
+    if (current_mask_state_ != mask_state) {
+      current_mask_state_ = mask_state;
+      drm_atomic_intf_->Perform(DRMOps::CONNECTOR_SET_FINGERPRINT_MASK,
+          token_.conn_id, current_mask_state_);
+      DLOGV("Display:%d Setting Fingerprint inDisplay Layer property = %d",
+            display_id_, current_mask_state_);
+    }
+  }
+#endif
+
   if (hw_panel_info_.partial_update && update_config) {
     if (IsFullFrameUpdate(*hw_layers_info)) {
       ResetROI();
diff --git a/sdm/libs/core/drm/hw_device_drm.h b/sdm/libs/core/drm/hw_device_drm.h
index 5fc3dd6875..613a9f85ba 100644
--- a/sdm/libs/core/drm/hw_device_drm.h
+++ b/sdm/libs/core/drm/hw_device_drm.h
@@ -361,6 +361,9 @@ class HWDeviceDRM : public HWInterface {
   bool autorefresh_ = false;
   std::unique_ptr<HWColorManagerDrm> hw_color_mgr_ = {};
   bool seamless_mode_switch_ = false;
+#ifdef CONNECTOR_PROP_UDFPS
+  uint32_t current_mask_state_ = 0;
+#endif
 };
 
 }  // namespace sdm
-- 
2.43.0

